torch/jit/__init__.py

```diff
@@ -1083,10 +1112,90 @@ def define(self, lang):
             # we add 1 to get to the proper surrounding scope.
             rcb = createResolutionCallback(frames_up=1)
             self._define(lang, rcb, True)
+
+    class WeakScriptModuleProxy(ScriptModule):
+        def __init__(self, original, stubs):
+            self.__dict__['_initialized'] = False
+            super(WeakScriptModuleProxy, self).__init__()
+
+            self.__dict__["_original"] = weakref.ref(original)
+            self.__dict__["_constants_set"] = set(getattr(original, "__constants__", []))
+
+            # Copy Parameters / Modules / Buffers
+            for name in dir(original):
+                item = getattr(original, name)
+                if isinstance(item, Parameter) or (isinstance(item, Module) and item is not self):
+                    ScriptModule.__setattr__(self, name, item)
+            for name in original._buffers:
+                self.register_buffer(name, original._buffers[name])
+
+            self.__dict__["_initialized"] = True
+            _create_methods_from_stubs(self, stubs)
+
+        def __getattr__(self, attr):
+            try:
+                return ScriptModule.__getattr__(self, attr)
+            except AttributeError:
+                if self.__dict__["_initialized"]:
+                    return getattr(self.__dict__["_original"](), attr)
+                else:
+                    raise AttributeError("Weak module has no attribute '{}'"
+                                         .format(attr))
+
+        def __setattr__(self, attr, value):
+            if not self.__dict__["_initialized"]:
+                return ScriptModule.__setattr__(self, attr, value)
+            if hasattr(self, attr):
+                return ScriptModule.__setattr__(self, attr, value)
+            else:
+                raise AttributeError("Cannot set new attribute '{}' on "
+                                     "weak script module once it has been "
+                                     "created".format(attr))
+
 else:
     ScriptModule = torch.nn.Module
 
 
+WeakModuleInstance = namedtuple('WeakModuleInstance', ('script_module', 'method_stubs', 'status'))
+WeakModule = namedtuple('WeakModule', ('instances', 'stubs'))
+
+
+def weak_module(cls):
+    if _enabled:
+        if weak_modules.get(cls) is None:
+            weak_modules[cls] = WeakModule(None, None)
+    return cls
+
+
+def _get_weak_stubs(mod):
```

Comment:
this function should take the class, not the module. It immediately does `mod.__class__` and then never uses mod